File: src/main/java/gregtechlite/gtlitecore/client/renderer/handler/PerspectiveAwareItemRenderer.java
package gregtechlite.gtlitecore.client.renderer.handler;

import codechicken.lib.render.item.IItemRenderer;
import com.morphismmc.morphismlib.collection.ListOps;
import gregtechlite.gtlitecore.client.renderer.EntityCallback;
import gregtechlite.gtlitecore.client.renderer.EntityItemTickCallback;
import net.minecraft.client.renderer.block.model.IBakedModel;
import net.minecraft.client.renderer.block.model.ItemOverrideList;
import net.minecraft.entity.EntityLivingBase;
import net.minecraft.entity.item.EntityItem;
import net.minecraft.item.ItemStack;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.World;
import net.minecraftforge.common.model.IModelState;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public abstract class PerspectiveAwareItemRenderer implements IItemRenderer, EntityItemTickCallback
{

    @Nullable
    protected EntityLivingBase renderEntity;
    protected World world;
    protected BlockPos entityPos;

    private final IModelState state;

    protected PerspectiveAwareItemRenderer(IModelState state)
    {
        this.state = state;
    }

    @Override
    public boolean isAmbientOcclusion()
    {
        return true;
    }

    @Override
    public boolean isGui3d()
    {
        return false;
    }

    @Override
    public ItemOverrideList getOverrides()
    {
        return new EntityCachingOverrideList((entity, world) -> {
            this.renderEntity = entity;
            this.world = world;
            if (entity != null)
                this.entityPos = entity.getPosition();
        });
    }

    @Override
    public IModelState getTransforms()
    {
        return this.state;
    }

    @Override
    public void onEntityTick(EntityItem item)
    {
        this.entityPos = item.getPosition();
    }

    private static class EntityCachingOverrideList extends ItemOverrideList
    {

        private final EntityCallback callback;

        public EntityCachingOverrideList(EntityCallback callback) {
            super(ListOps.of());
            this.callback = callback;
        }

        @NotNull
        @Override
        public IBakedModel handleItemState(@NotNull IBakedModel originalModel,
                                           @NotNull ItemStack stack,
                                           World world, EntityLivingBase entity)
        {
            this.callback.onEntityStuffs(entity, world);
            return super.handleItemState(originalModel, stack, world, entity);
        }

    }

}

File: src/main/kotlin/gregtechlite/gtlitecore/client/renderer/handler/PerspectiveAwareItemRenderer.kt
package gregtechlite.gtlitecore.client.renderer.handler

import codechicken.lib.render.item.IItemRenderer
import gregtechlite.gtlitecore.client.renderer.EntityCallback
import gregtechlite.gtlitecore.client.renderer.EntityItemTickCallback
import net.minecraft.client.renderer.block.model.IBakedModel
import net.minecraft.client.renderer.block.model.ItemOverrideList
import net.minecraft.entity.EntityLivingBase
import net.minecraft.entity.item.EntityItem
import net.minecraft.item.ItemStack
import net.minecraft.util.math.BlockPos
import net.minecraft.world.World
import net.minecraftforge.common.model.IModelState

abstract class PerspectiveAwareItemRenderer protected constructor(private val state: IModelState?) : IItemRenderer, EntityItemTickCallback
{

    @JvmField
    protected var renderEntity: EntityLivingBase? = null
    @JvmField
    protected var world: World? = null
    @JvmField
    protected var entityPos: BlockPos? = null

    override fun isAmbientOcclusion(): Boolean = true

    override fun isGui3d(): Boolean = false

    override fun getOverrides(): ItemOverrideList
    {
        return EntityCachingOverrideList { entity, world ->
            this.renderEntity = entity
            this.world = world
            if (entity != null) this.entityPos = entity.position
        }
    }

    override fun getTransforms(): IModelState? = this.state

    override fun onEntityTick(item: EntityItem?)
    {
        this.entityPos = item?.position
    }

    private class EntityCachingOverrideList(private val callback: EntityCallback) : ItemOverrideList(listOf())
    {

        override fun handleItemState(originalModel: IBakedModel,
                                     stack: ItemStack,
                                     world: World?,
                                     entity: EntityLivingBase?): IBakedModel
        {
            this.callback.onEntityStuffs(entity, world)
            return super.handleItemState(originalModel, stack, world, entity)
        }

    }

}